[linux-6.6.y] iommu/vt-d: Add support for detecting ACPI namespace device in RMRR

leoliu-oc opened the pull request:
zhaoxin inclusion
category: feature
CVE: NA

-----------------

As below, ZX-200 xHCI mcu is a RMRR ANDD device in some case.

[060h 0096   2]                Subtable Type : 0001 [Reserved Memory Region
[062h 0098   2]                       Length : 0020

[064h 0100   2]                     Reserved : 0000
[066h 0102   2]           PCI Segment Number : 0000
[068h 0104   8]                 Base Address : 00000000B5DA5000
[070h 0112   8]          End Address (limit) : 00000000B5DDDFFF

[078h 0120   1]            Device Scope Type : 05 [Namespace Device]
[079h 0121   1]                 Entry Length : 08
[07Ah 0122   2]                     Reserved : 0000
[07Ch 0124   1]               Enumeration ID : 02
[07Dh 0125   1]               PCI Bus Number : 09

[07Eh 0126   2]                     PCI Path : 12,00

iommu driver cannot find this device and build identity map for the RMRR region, DMAR faults would occur for xHCI controller.

Add func dmar_acpi_bus_add_dev to find the RMRR ANDD device.

Add func acpi_rmrr_andd_probe to build identity map for the RMRR region into the domain of the correspanding xHCI controller.

Add func iova_reserve_domain_addr to keep away from RMRR region when using dma iova.

opsiff commented on drivers/iommu/intel/iommu.c:
```diff
@@ -3725,6 +3743,43 @@ static int __init platform_optin_force_iommu(void)
+
+	ret = iommu_probe_device(dev);
+
+	iommu = device_lookup_iommu(dev, &bus, &devfn);
```
这里有报错 是不是使用了没定义的 device_lookup_iommu 函数?
drivers/iommu/intel/iommu.c: In function ‘acpi_rmrr_andd_probe’:
drivers/iommu/intel/iommu.c:3766:17: error: implicit declaration of function ‘device_lookup_iommu’; did you mean ‘device_to_iommu’? [-Werror=implicit-function-declaration]
3766 |         iommu = device_lookup_iommu(dev, &bus, &devfn);
     |                 ^~~~~~~~~~~~~~~~~~~
     |                 device_to_iommu
drivers/iommu/intel/iommu.c:3766:15: error: assignment to ‘struct intel_iommu *’ from ‘int’ makes pointer from integer without a cast [-Werror=int-conversion]
3766 |         iommu = device_lookup_iommu(dev, &bus, &devfn);

leoliu-oc replied:
是的，这边测试时使用的是主线LTS-6.6分支源码，在最新主线源码中，device_to_iommu更新为了device_lookup_iommu。这边会修正后再提交。

leoliu-oc committed:
iommu/vt-d: Add support for detecting ACPI namespace device in RMRR

zhaoxin inclusion
category: feature
CVE: NA

-----------------

As below, ZX-200 xHCI mcu is a RMRR ANDD device in some case.

[060h 0096   2]                Subtable Type : 0001 [Reserved Memory Region
[062h 0098   2]                       Length : 0020

[064h 0100   2]                     Reserved : 0000
[066h 0102   2]           PCI Segment Number : 0000
[068h 0104   8]                 Base Address : 00000000B5DA5000
[070h 0112   8]          End Address (limit) : 00000000B5DDDFFF

[078h 0120   1]            Device Scope Type : 05 [Namespace Device]
[079h 0121   1]                 Entry Length : 08
[07Ah 0122   2]                     Reserved : 0000
[07Ch 0124   1]               Enumeration ID : 02
[07Dh 0125   1]               PCI Bus Number : 09

[07Eh 0126   2]                     PCI Path : 12,00

iommu driver cannot find this device and build identity map for the RMRR
region, DMAR faults would occur for xHCI controller.

Add func dmar_acpi_bus_add_dev to find the RMRR ANDD device.

Add func acpi_rmrr_andd_probe to build identity map for the RMRR region
into the domain of the correspanding xHCI controller.

Add func iova_reserve_domain_addr to keep away from RMRR region when using
dma iova.

Signed-off-by: leoliu-oc <leoliu-oc@zhaoxin.com>

opsiff commented:
/ok-to-test
